File: problems-originals/2801-2900/2831_find-the-longest-equal-subarray/solution.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestEqualSubarray = function (nums, k) {
    const positionsByValue = new Map();
    for (let i = 0; i < nums.length; i++) {
        const list = positionsByValue.get(nums[i]);
        if (list === undefined) {
            positionsByValue.set(nums[i], [i]);
        } else {
            list.push(i);
        }
    }
    let answer = 0;
    for (const positions of positionsByValue.values()) {
        let left = 0;
        for (let right = 0; right < positions.length; right++) {
            // Span length minus kept copies is the deletion cost.
            while (positions[right] - positions[left] - (right - left) > k) {
                left++;
            }
            answer = Math.max(answer, right - left + 1);
        }
    }
    return answer;
};

File: problems-originals/2801-2900/2831_find-the-longest-equal-subarray/solution.py
from typing import List


class Solution:
    def longestEqualSubarray(self, nums: List[int], k: int) -> int:
        positions_by_value: dict[int, List[int]] = {}
        for i, value in enumerate(nums):
            positions_by_value.setdefault(value, []).append(i)
        answer = 0
        for positions in positions_by_value.values():
            left = 0
            for right in range(len(positions)):
                # Span length minus kept copies is the deletion cost.
                while (positions[right] - positions[left]) - (right - left) > k:
                    left += 1
                answer = max(answer, right - left + 1)
        return answer
